src/kv/value_wrapper.hpp

```diff
@@ -0,0 +1,183 @@
+/**
+ * Canary - A free and open-source MMORPG server emulator
+ * Copyright (©) 2019-2022 OpenTibiaBR <opentibiabr@outlook.com>
+ * Repository: https://github.com/opentibiabr/canary
+ * License: https://github.com/opentibiabr/canary/blob/main/LICENSE
+ * Contributors: https://github.com/opentibiabr/canary/graphs/contributors
+ * Website: https://docs.opentibiabr.com/
+ */
+
+#pragma once
+
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <variant>
+#include <optional>
+#include <algorithm>
+#include <ranges>
+#include <iterator>
+#include <type_traits>
+
+class ValueWrapper;
+
+using StringType = std::string;
+using IntType = int;
+using DoubleType = double;
+using ArrayType = std::vector<ValueWrapper>;
+using MapType = phmap::flat_hash_map<std::string, std::shared_ptr<ValueWrapper>>;
+
+using ValueVariant = std::variant<StringType, IntType, DoubleType, ArrayType, MapType>;
+
+class ValueWrapper {
+public:
+	explicit ValueWrapper(uint64_t timestamp = 0);
+	explicit(false) ValueWrapper(const ValueVariant &value, uint64_t timestamp = 0);
+	explicit(false) ValueWrapper(const std::string &value, uint64_t timestamp = 0);
+	explicit(false) ValueWrapper(int value, uint64_t timestamp = 0);
+	explicit(false) ValueWrapper(double value, uint64_t timestamp = 0);
```

Comment:
I think that you can template those, except for the map/list ones.

template <typename T>
ValueWrapper::ValueWrapper(T value, uint64_t timestamp) : 
    data_(value), timestamp_(timestamp) { }

Reply:
good point -- I was trying earlier to let the implicit conversion take over but that doesn't work. Template should.

Reply:
Humm, actually, tried this it gets the `<<` confused

Reply:
Interesting, well not a blocker, but would be a nice clean up.

Reply:
gonna punt on this for now, marking as resolved and I'll look into it after this is in main